spec/index.bs

```diff
@@ -2883,57 +2964,79 @@ The {{GPUMapMode}} flags determine how a {{GPUBuffer}} is mapped when calling
+                        - |this| is a [=valid=] {{GPUBuffer}}.
+
+                            Issue: Figure out exactly how [=invalid=] and destroyed buffers are handled.
+                        - |offset| is a multiple of 8.
+                        - |rangeSize| is a multiple of 4.
+                        - |offset| + |rangeSize| &le; |this|.{{GPUBuffer/size}}
+                        - |this|.{{GPUBuffer/[[state]]}} is [=buffer state/unmapped=]
+                        - |mode| contains exactly one of {{GPUMapMode/READ}} or {{GPUMapMode/WRITE}}.
+                        - If |mode| contains {{GPUMapMode/READ}} then |this|.{{GPUBuffer/usage}} must contain {{GPUBufferUsage/MAP_READ}}.
+                        - If |mode| contains {{GPUMapMode/WRITE}} then |this|.{{GPUBuffer/usage}} must contain {{GPUBufferUsage/MAP_WRITE}}.
+
+                        Issue: Do we validate that |mode| contains only valid flags?
+                    </div>
 
-                Then:
+                    Then:
 
-                1. Record a validation error on the current scope.
-                1. Return [=a promise rejected with=] an {{OperationError}} on the [=Device timeline=].
+                    1. Issue the following steps on |contentTimeline|.
 
-            1. Let |p| be a new {{Promise}}.
-            1. Set |this|.{{GPUBuffer/[[mapping]]}} to |p|.
-            1. Set |this|.{{GPUBuffer/[[state]]}} to [=buffer state/mapping pending=].
-            1. Set |this|.{{GPUBuffer/[[map_mode]]}} to |mode|.
-            1. Enqueue an operation on the default queue's [=Queue timeline=] that will execute the following:
+                        <ol data-timeline=content>
+                            1. [=Reject=] |p| with an {{OperationError}}.
+                        </ol>
 
-                <div class=queue-timeline>
-                    1. If |this|.{{GPUBuffer/[[state]]}} is [=buffer state/mapping pending=]:
+                    1. [$Generate a validation error$] and stop.
 
-                        1. Let |m| be a new {{ArrayBuffer}} of size |rangeSize|.
+                1. Once all <span data-timeline=queue>currently-enqueued operations (on any queue)</span>
+                    on |this| are completed, issue the |device timeline steps to map| on this timeline.
+            </ol>
 
-                            Note: This is just a memory allocation; this {{ArrayBuffer}} object is not exposed directly.
-                        1. Set the content of |m| to the content of |this|'s allocation starting at offset |offset| and for |rangeSize| bytes.
-                        1. Set |this|.{{GPUBuffer/[[mapping]]}} to |m|.
-                        1. Set |this|.{{GPUBuffer/[[state]]}} to [=buffer state/mapped=].
-                        1. Set |this|.{{GPUBuffer/[[mapping_range]]}} to <code>[|offset|, |offset| + |rangeSize|]</code>.
-                        1. Set |this|.{{GPUBuffer/[[mapped_ranges]]}} to `[]`.
+            The |device timeline steps to map| are:
```

Comment:
Not clear to me why these two sets of steps have been separated, given that they're on the same timeline. Adds to the difficulty reading this section that I mentioned in other comments. It would be good to come up with a consistent way of stating "Wait for async operation X to complete." without implying that the entire timeline is blocked until that happens so that steps like this can be written more concisely.

Reply:
I actually like this separation. It makes each of these blocks effectively one C++ function, a block of stuff that always happens all at once. "Do steps later" becomes explicit. Writing this an a JS-async style would probably be fine though.

I've updated the style to be a bit more concise.

Also, if it turns out we don't actually have to do anything in the second device timeline box, we can get rid of it.